### ISSUE TITLE
Add an ability to use soft assert with shouldBe etc.

### ISSUE THREAD

**Issue by sytnykma-ss** [selenide/selenide#1453]
Please add an ability to use soft assert with selenide asserts like:
SelenideSoftAseert sa = new SelenideSoftAssert();
sa.$x("asdad").shouldBe(visible, etc); / $x("asdad").sa.shouldBe(visible);
sa.$x("1234").shouldBe(visible, etc); / $x("123").sa.shouldBe(visible);
sa.assertAll();
 
It will be really helpful to check a lot of ui elements states on the page at the same time instead of interrupting test execution if the first element checking fails but other elements right after it may be good.

**Comment by asolntsev** [comment]
Hi @sytnykma-ss !
Selenide already supports soft asserts, but with a different syntax. 

You just need to 
1. enable soft asserts: `Configuration.assertionMode = SOFT;`
2. add rule (for JUnit4) or extension (for JUnit5) or listener (for TestNG) called `SoftAsserts`
- and nothing more. 

Starting from this moment, all checks like `$.shouldBe(visible);` will be soft.

**Comment by sytnykma-ss** [comment]
But I dont need the soft checks for a whole project, I just want to use it in exact case or test.

**Comment by asolntsev** [comment]
@sytnykma-ss currently we don't have such a feature. Probably we should implement it - it should be quite simple.

**Comment by sytnykma-ss** [comment]
Yep, Im asking to add this feature and you said me to put the suggestion here =).
Because I don't want my test method to continue executing if current assertion fails. Like:
@TmsLink("COM-41642")
    @Test
    public void checkMessageUi() {
        contactPage.getSendTextMessagePopUp().getSendMessageBtn().shouldBe(visible, disabled);
        contactPage.getSendTextMessagePopUp().getTextInput().shouldBe(visible).sendKeys("Some text");
        contactPage.getSendTextMessagePopUp().getCalculateCostBtn().shouldBe(visible, enabled).click();
        contactPage.getSendTextMessagePopUp().getCancelBtn().shouldBe(visible);
        contactPage.getSendTextMessagePopUp().getPrimaryPhoneNumber().shouldBe(visible);
    }
I want to check all of the elements at the same time (better to use soft assert from selenide here). 
But here
@Test()
    public void checkUiDesign() {
        emailsReportPage.getUpgradePlanBtn().shouldHave(appear, enabled);
        EmailOverviewFragment emailOverviewFragment = emailsReportPage.getEmailOverviewFragment();
        etc....
    }
I dont need to continue execution if emailsReportPage.getUpgradePlanBtn().shouldHave(appear, enabled); failed to not waste time

**Comment by asolntsev** [comment]
@sytnykma-ss I tried to implement this feature and found that it's not trivial. I need more time. :(

**Comment by zhupengfarmer** [comment]
> @sytnykma-ss我试图实现这个功能，发现它不是微不足道的。我需要更多的时间。:(

element.exists()   Don't wait, judge directly, but don't interrupt the test process
element.shouldHave(Condition.exist)    wait, judge directly, but interrupt the test process

Sometimes I need to determine whether the element appears in Configuration.timeout. Can it be added? Wait in Configuration.timeout to determine exist

i don't want to interrupt the testing process

**Comment by vinogradoff** [comment]
@sytnykma-ss there is an option for that, which may or may not have some minmal side-effects for multi-threaded execution:

You can either put in a single test:

	  Configuration.assertionMode = SOFT;
	  ....
	  	
	  Configuration.assertionMode = STRICT;

or put in every test class in BeforeAll
      Configuration.assertionMode = STRICT;

and add just for test where you need it:

	  Configuration.assertionMode = SOFT;

**Comment by DenisKonoplyankoArammeem** [comment]
Hi! Are there any plans to develop this feature? This would be very helpful for projects with multiple threads.

Currently, Configuration.assertionMode property affects whole project and softAsserts cannot be used is specific places (tests)

**Comment by asolntsev** [comment]
@DenisKonoplyankoArammeem as we discussed above, we are agree this feature might be implemented. But I personally don't have it in my top priorities.  You are free to make your contribution and submit a pull request. ;)

**Comment by itkhanz** [comment]
Hi,
I have a similar usecase where I need soft asserts only in specific test cases and blocks but as far as I know Selenide offers Soft Asserts that can only be applied at the project level and not individual methods or code blocks within methods.

It would be nice to have this feature in newer releases of Selenide with screenshots captured automatically for each of the soft assertion that failed.

I came up with a workaround to combine the usage of soft assertions from TestNG with the SelenideElement. Although it is [not recommended to use getter methods](https://github.com/selenide/selenide/wiki/do-not-use-getters-in-tests) like is() or has() but I added custom timeouts as extra parameters to make sure that the element is actually visible or present in the DOM before returning the value.
Here is how the solution looks like: 

```
import org.testng.annotations.Test;
import org.testng.asserts.SoftAssert;

import java.time.Duration;

import static com.codeborne.selenide.Selenide.*;
import static com.codeborne.selenide.Condition.*;

public class Example {

    @Test
    public void workaround_for__SoftAssert_Selenide_TestNG_Test() {
        open("https://the-internet.herokuapp.com/");

        String title = $("h1").shouldBe(visible).text();
        String firstLinkText = $$("ul li a").first().shouldBe(visible).getAttribute("href");
        boolean imgAltAttrPresence = $("img").shouldBe(exist).has(attribute("alt"), Duration.ofSeconds(10));
        boolean footerVisibility = $("#page-footer").is(visible, Duration.ofSeconds(10));

        SoftAssert softAssert = new SoftAssert();
        softAssert.assertEquals(title, "Welcome to the-internet");
        softAssert.assertEquals(firstLinkText, "https://the-internet.herokuapp.com/abtest");
        softAssert.assertTrue(imgAltAttrPresence);
        softAssert.assertTrue(footerVisibility);
        softAssert.assertAll();

    }
}
```

This probably solves the issue of waiting for some condition but it will not capture the screenshot automatically I guess as the assertion is being made by TestNG and not the Selenide. 

@asolntsev  I would appreciate your inputs on this hacky solution. What could possibly go wrong and any suggestions for improvements?

**Comment by asolntsev** [comment]
> your inputs on this hacky solution

@itkhanz No, you solution is not hacky. I think it works (if the web elements are not reloading etc. - this is the reason why getters are not recommended). 

You can even add `ScreenShooter` extension to automatically take a screenshot in case of test failure: https://selenide.org/documentation/screenshots.html

**Comment by Merckan** [comment]
If you're using **JUnit 5**, you can implement something like the following:

In your **Page** class:

```
@SafeVarargs
public final YourPage shouldAll(Consumer<YourPage>... consumers) {
    var executables = Arrays.stream(consumers)
            .<Executable>map(consumer -> () -> consumer.accept(this))
            .toList();
    Assertions.assertAll(executables);
    return this;
}
```

In your **test**:

```
new YourPage()
        .shouldAll(
                page -> page.someInputShould(have(ownText("expected text"))),
                page -> page.someCheckboxShould(exist)
        );
```

**Note[!]:** Each Selenide assertion will use the default or explicitly specified timeout.
For example, if the default timeout is 10 seconds, and you have 10 soft assertions that all fail, the total execution time before the test fails could reach 10 × 10 = 100 seconds. Keep this in mind when writing multiple soft assertions.
